### PR TITLE
Update OkHttp to 2.3.0, kill ConnectionWrapper

### PR DESCRIPTION
**Pull request by aried3r**
This supersedes #1204.

### CODE REVIEW

**Comment by aried3r** [comment]
I'm not quite sure why the test run failed, I might be missing something.

/cc @robertpainsi @obusher

**Comment by robertpainsi** [comment]
Oh, didn't see the mentioning in the first place, sy
I guess @SimonStefan is already on it.

JFI:
The emulators didn't start up
```
23:06:23 [android] Emulator did not appear to start; giving up
```
And the device throw an error
```
23:06:28 ERROR: Publisher hudson.tasks.junit.JUnitResultArchiver aborted due to exception
23:06:28 hudson.AbortException: No test report files were found. Configuration error?
```

**Comment by robertpainsi** [comment]
ok to test

**Comment by robertpainsi** [comment]
retest please

**Comment by obusher** [comment]
The testing device error looks like [this](https://jira.catrob.at/browse/JENKINS-49). :tangerine:

**Comment by SimonStefan** [comment]
Yes, I was looking at the code it seems fine. But I don't have time to test it on a device using some edge cases like no internet connection,...
Could anybody else do this?

**Comment by robertpainsi** [comment]
@aried3r: There was a [YellowJenkins build](https://jenkins.catrob.at/job/Catroid-PullRequest-TestJob/186/) of your PR with just two failures (maybe related to your changes (didn't check)):

```
junit.framework.AssertionFailedError: Upload did work, but error toastmessage should have been displayed
at org.catrobat.catroid.uitest.web.ProjectUpAndDownloadTest.testUploadProjectOffensiveLanguageUsed(ProjectUpAndDownloadTest.java:190)
```

```
junit.framework.AssertionFailedError: Upload did work, but error toastmessage should have been displayed
at org.catrobat.catroid.uitest.web.ProjectUpAndDownloadTest.testUploadProjectOldCatrobatLanguageVersion(ProjectUpAndDownloadTest.java:158)
```

File: catroid/src/org/catrobat/catroid/web/ServerCalls.java [hunk 1]
```diff
@@ -82,102 +103,162 @@
...
+					.addFormDataPart(
+							Constants.TOKEN,
+							token)
+					.addFormDataPart(
+							Constants.USERNAME,
+							username)
+					.addFormDataPart(
+							DEVICE_LANGUAGE,
+							language)
+					.build();
```
Comment (aried3r) [inline catroid/src/org/catrobat/catroid/web/ServerCalls.java]: Maybe I'll go ahead and explain a few changes. I killed the `ConnectionWrapper`, because it was another level of indirection that – in my opinion – did not add any value. Also it was hard to see _what_ exactly was being POSTed, because things kept being added to a map before it was actually used.

The way I formatted the code is of course very verbose, but I'd like to think of it as being easily readable. I think it's a lot better than before, but still has room for improvement.

File: catroid/src/org/catrobat/catroid/web/ServerCalls.java [hunk 2]
```diff
@@ -82,102 +103,162 @@
...
-	private ConnectionWrapper connection;
 	private String emailForUiTests;
-	private int uploadStatusCode;
+	private final OkHttpClient okHttpClient;
+	private final Gson gson;
 
 	private ServerCalls() {
-		connection = new ConnectionWrapper();
+		okHttpClient = new OkHttpClient();
+		okHttpClient.setConnectionSpecs(Arrays.asList(ConnectionSpec.MODERN_TLS));
```
Comment (aried3r) [inline catroid/src/org/catrobat/catroid/web/ServerCalls.java]: This means that cleartext connections are not allowed. You'd have to add `ConnectionSpec.CLEARTEXT` if you want that.

File: catroid/src/org/catrobat/catroid/web/ServerCalls.java [hunk 3]
```diff
@@ -82,102 +103,162 @@
...
+		gson = new Gson();
```
Comment (aried3r) [inline catroid/src/org/catrobat/catroid/web/ServerCalls.java]: Gson, because parsing JSON manually is just crazy, give me an object already! ;)

File: catroidTest/src/org/catrobat/catroid/uitest/web/ProjectUpAndDownloadTest.java [hunk 4]
```diff
@@ -193,14 +187,7 @@ public void testUploadProjectOffensiveLanguageUsed() throws Throwable {
 
 		boolean uploadErrorOccurred = solo.waitForText(solo.getString(R.string.error_project_upload));
 
-		int statusCode = 0;
-		int statusCodeOffensiveLanguage = 511;
-		statusCode = (Integer) Reflection.getPrivateField(ServerCalls.getInstance(), "uploadStatusCode");
```
Comment (aried3r) [inline catroidTest/src/org/catrobat/catroid/uitest/web/ProjectUpAndDownloadTest.java]: Reflection removed, because I don't think a UI test should rely on a private field of a class it doesn't even know about.

**Comment by aried3r** [comment]
I'll have a look at the tests that failed in the coming days.

**Comment by avihric** [comment]
Any news on this @aried3r and what about #1204 ?

**Comment by aried3r** [comment]
I'll try to get this fixed and merged before you build a new release version. When is the deadline for merges?

**Comment by aried3r** [comment]
Hopefully [this testrun](https://jenkins.catrob.at/view/All-Categories/view/Catroid-multi-job/job/Catroid-Multi-Job-Custom-Branch-RELOADED/1529/) will be green, but I think there's quite some traffic going on on Jenkins. I won't be at my PC in the next few hours, please drop me a message here if there's anything I can do.

**Comment by obusher** [comment]
ok to test

**Comment by obusher** [comment]
ok to test

**Comment by avihric** [comment]
@aried3r Could you open this PR for the release branch please

**Comment by aried3r** [comment]
Done, see #1227.
